Add phase space diagnostics

tom48701 opened the pull request:
### Ad hoc phase space diagnostics for fbpic
Just a temporary measure until there is an official implementation

The core of it is the `numpy.histogramdd` function, and the resulting dataset is appended to the usual hdf5 output files. The axis limits, names and quanaity desposited are saved as attributes of the dataset.
The dataset of the histogram is located within the diag file at

`data/[iteration]/phase spaces/[species]/[diagnostic name]`

with the attributes

`quantity` : deposited quantity
`axis1` : (min, max)
`axis1Name` : axis name

with additional axes numbered sequentially in the order they are passed.

Currently each instance of the `PhaseSpaceDiagnostic` class describes one diagnostic. There's no reason why it couldn't be set up to process multiple diagnostics from a single instance, and in fact I wrote it that way initially, but in practice it very quickly gets messy passing long lists of arguments defining all aspects of the required diagnostics. 
I found it easier to keep track of things just defining a new instance for each diagnostic, So for the sake of keeping things simple this is what I settled on. (Passing multiple species at once still works though)

To do:

Tidy up the metadata. Since it's adapted from an existing class for a different type of diagnostic, the metadata generated is lacking in some areas and redundant in others.

RemiLehe commented on fbpic/openpmd_diag/phasespace_diag.py:
```diff
@@ -0,0 +1,540 @@
+# Copyright 2016, FBPIC contributors
+# Authors: Remi Lehe, Manuel Kirchen
```
```suggestion
# Copyright 2020, FBPIC contributors
# Authors: Thomas Wilson, Remi Lehe, Manuel Kirchen
```

RemiLehe commented on fbpic/openpmd_diag/phasespace_diag.py:
```diff
@@ -0,0 +1,540 @@
+            dset.attrs['deposit'] = 'weight'        
+  
+        # generate the histogram
+        hist, edges = np.histogramdd(sample, bins=self.bins, 
```
I would suggest to also put this call to `np.histogramdd` in the `if self.rank == 0:` condition below:
only the MPI rank that writes the results needs to perform this calculation.

RemiLehe commented on fbpic/openpmd_diag/phasespace_diag.py:
```diff
@@ -0,0 +1,540 @@
+            dset[:] = hist
+            for i in range(self.Ndims):
+                dset.attrs['axis%i'%(i+1)] = [edges[i][0], edges[i][-1]]
+                dset.attrs['axis%iName'%(i+1)] = '%s'%self.phase_space[i]
```
As mentioned above, I would suggest to replace the above meta-data by `axisLabel`, `gridSpacing`, `gridGlobalOffset`.

RemiLehe reviewed (commented):
Thanks a lot for contributing this to FBPIC!
The code is well written and easy to understand. I have a few comments and suggestions, mainly related to the metadata. (see the inline comments and suggestions).

In order to update the current pull request accordingly, you can implement the changes on your local computer and then push them to the same branch, with `git`.

Thanks again!

tom48701 commented:
I have implemented the attribute changes, as you suggested, now the attributes for an N-dimensional dataset are:

`gridUnitsSI` an array of length N
`gridUnitDimension` an array of length 7*N (as per https://github.com/openPMD/openPMD-standard/pull/193)
`axisLabels` a list of length N containing the axis names
`gridSpacing` and `gridGlobalOffset` are analagous to those in a field diagnostic

the only attributes that don't seem to have analagous terms in other diagnostics are those describing the deposited quantity, so this information is contained in the

`depositedQuantityUnit`, `depositedQuantityUnitSI` and `depositedQuantityUnitDimension` attributes

these can be easily changed or removed if needed.

In addition to these changes, I have also added an additional feature.
There is a new class; `CustomQuantity`, which contains instructions to compute a custom particle quantity from the 'basic' particle quantities.

Currently it is just a simple collection of attributes describing the name of the quantity for reference purposes, a function describing how to compute the quantity, a list of neccessary arguments to that function and the dimensions of the computed quantity.

A list of instances of this class can be passed to the `PhaseSpaceDiagnostic`, and these will be taken into consideration when performing the diagnostic.

For instance, the radial position of a particle is given by sqrt(x^2 + y^2). A `CustomQuantity` class instance to compute this would be initialised as follows:

```
def r_func(args):
	return( np.sqrt(args[0]**2+args[1]**2) )

CustomQuantity('r', r_func, ('x','y'), [1.,0.,0.,0.,0.,0.,0.])
```

Passing this to the diagnostic would open up the 'r' quantity to be used as either an axis or a deposited quantity. So far I have tested this with angular momentum and the particle pitch angle and it seems both functional and useful.

comments:

In the attributes, only the short name of the quantity is deposited, i.e. 'w' instead of 'weighting'. Maybe this should be changed so the full names are written for clarity?
I'm still not sure exactly what is required and what can be omitted from the metadata, so that probably still needs work
